eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/ChannelRelay.scala

```diff
@@ -112,29 +115,65 @@ class ChannelRelay private(nodeParams: NodeParams,
   private val forwardFailureAdapter = context.messageAdapter[Register.ForwardFailure[CMD_ADD_HTLC]](WrappedForwardFailure)
   private val addResponseAdapter = context.messageAdapter[CommandResponse[CMD_ADD_HTLC]](WrappedAddResponse)
 
+  private val nextNodeId_opt = r.payload.outgoing match {
+    case Left(walletNodeId) => Some(walletNodeId)
+    // All the channels point to the same next node, we take the first one.
+    case Right(_) => channels.headOption.map(_._2.nextNodeId)
+  }
+
+  /** Channel id explicitly requested in the onion payload. */
+  private val requestedChannelId_opt = r.payload.outgoing match {
+    case Left(_) => None
+    case Right(outgoingChannelId) => channels.collectFirst {
+      case (channelId, channel) if channel.shortIds.localAlias == outgoingChannelId => channelId
+      case (channelId, channel) if channel.shortIds.real.toOption.contains(outgoingChannelId) => channelId
+    }
+  }
+
   private case class PreviouslyTried(channelId: ByteVector32, failure: RES_ADD_FAILED[ChannelException])
 
-  def relay(previousFailures: Seq[PreviouslyTried]): Behavior[Command] = {
+  def start(): Behavior[Command] = {
+    r.payload.outgoing match {
+      case Left(walletNodeId) => wakeUp(walletNodeId)
+      case Right(requestedShortChannelId) => relay(Some(requestedShortChannelId), Seq.empty)
```

Comment:
Note how this can be extended to detect "custom" scids and insert the wake-up step.